### PR TITLE
feat: unify mayastor-client address/port flags

### LINKED ISSUE #707: mayastor-client has subtle way to point it at a mayastor
The `mayastor-client` binary has `-a` and `-p` options for choosing an address and port of the target mayastor:

<pre>$ mayastor-client
 Mayastor CLI 0.1
 CLI utility for Mayastor

USAGE:
 mayastor-client <span class="error">[FLAGS]</span> <span class="error">[OPTIONS]</span>

FLAGS:
 -h, --help Prints help information
 -q, --quiet Do not print any output except

 <span class="code-object">for</span> list records
 -V, --version Prints version information
 -v, --verbose Verbose output

OPTIONS:
 -a, --address

 <host>

 IP address of mayastor instance <span class="error">[default: 127.0.0.1]</span>
 -p, --port 

 <number>

 Port number of mayastor server <span class="error">[default: 10124]</span>
 -u, --units 

 <base>
 Output with large units: i 

 <span class="code-object">for</span> kiB, etc. or d 

 <span class="code-object">for</span> kB, etc.

SUBCOMMANDS:
 bdev Block device management
 device Host devices
 help Prints this message or the help of the given subcommand(s)
 jsonrpc Call a json-rpc method with a raw JSON payload
 nexus Nexus device management
 perf Performance statistics
 pool Storage pool management
 rebuild Rebuild management
 replica Replica management
 snapshot Snapshot management

 </number>

 </host>

</pre>

Unfortunately for users, these options aren't shown in global helps:

<pre>$ mayastor-client nexus
 mayastor-client-nexus 
 Nexus device management

USAGE:
 mayastor-client nexus

FLAGS:
 -h, --help Prints help information
 -V, --version Prints version information

SUBCOMMANDS:
 add add a child
 child Nexus child management
 children list nexus children
 create Create a new nexus device
 destroy destroy the nexus with given name
 help Prints this message or the help of the given subcommand(s)
 list list all nexus devices
 publish publish the nexus
 remove remove a child
 unpublish unpublish the nexus

</pre>

They're also not usable globally:

<pre>$ mayastor-client nexus -a 123.0.0.1
 error: Found argument '-a' which wasn't expected, or isn't valid in this context

USAGE:
 mayastor-client nexus

For more information try --help

</pre>

Probably the easiest method is via `clap::Arg::global` [https://docs.rs/clap/2.21.3/clap/struct.Arg.html#method.global](https://docs.rs/clap/2.21.3/clap/struct.Arg.html#method.global)

### PR DESCRIPTION
**Pull request by Hoverbear**
Unifies the `-a` and `-p` flags to a `-h` flag which
takes a full url (optionally) without a scheme.

Further, this makes the flag global, so it can be used
in any position:

```bash
RUST_BACKTRACE=1
RUST_LOG=mayastor=trace
cargo run -- nexus list
cargo run -- --bind 127.0.0.1 nexus list
cargo run -- -b 127.0.0.1:10124 nexus list
cargo run -- --bind http://127.0.0.1:10124 nexus list
cargo run -- --bind http://127.0.0.1 nexus list
cargo run -- nexus list --bind 127.0.0.1
cargo run -- nexus list --bind 127.0.0.1:10124
cargo run -- nexus list -b http://127.0.0.1:10124
cargo run -- nexus list --bind http://127.0.0.1
```

Closes #707.

### CODE REVIEW

**Comment by cjones1024** [comment]
Not sure about the choice of -h here which is generally reserved as an alias for --help.
A possibly  better solution is to still use -a but now with the new meaning - ie. allow the port to be included as part of the address.

**Comment by Hoverbear** [comment]
Agreed, will move it to `-a`... I was also thinking about `-b` for `--bind`.

**Comment by Hoverbear** [comment]
Ah, remember why I moved off `-a`, since this is moving to global, `-a` can be passed anywhere, and this looks like `--all` in some locations, indeed, it conflicts with `--all`.

I am going to move it to `bind` for you to consider.

File: mayastor/src/bin/mayastor-client/context.rs [hunk 1]
```diff
@@ -1,7 +1,35 @@
...
 use clap::ArgMatches;
-use std::cmp::max;
+use http::uri::{Scheme, Uri, Authority, PathAndQuery};
+use snafu::{Backtrace, ResultExt, Snafu};
+use std::{cmp::max, str::FromStr};
+use tonic::transport::Endpoint;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Invalid URI bytes: {}", source))]
```
Comment (tiagolobocastro) [inline mayastor/src/bin/mayastor-client/context.rs:12]: You don't need to add the source with snafu

File: mayastor/src/bin/mayastor-client/context.rs [hunk 2]
```diff
@@ -22,27 +50,46 @@ impl Context {
             .value_of("units")
             .and_then(|u| u.chars().next())
             .unwrap_or('b');
-        let endpoint = {
-            let addr = matches.value_of("address").unwrap_or("127.0.0.1");
-            let port = value_t!(matches.value_of("port"), u16).unwrap_or(10124);
-            format!("{}:{}", addr, port)
+        // Ensure the provided host is defaulted & normalized to what we expect.
+        // TODO: This can be significantly cleaned up when we update tonic 0.1 and its ilk.
```
Comment (tiagolobocastro) [inline mayastor/src/bin/mayastor-client/context.rs:54]: what's ilk?

**Review by tiagolobocastro — APPROVED** [review]
bind seems quite appropriate :+1:

File: mayastor/src/bin/mayastor-client/context.rs (hunk 2 above)
Comment (Hoverbear) [reply mayastor/src/bin/mayastor-client/context.rs:54]: "like", let me fix it...

**Review by jkryl — APPROVED** [review]

**Review by cjones1024 — APPROVED** [review]

**Comment by Hoverbear** [comment]
bors merge

**Comment by Hoverbear** [comment]
bors merge

File: mayastor/src/bin/mayastor-client/context.rs (hunk 1 above)
Comment (tiagolobocastro) [reply mayastor/src/bin/mayastor-client/context.rs]: Apologies I misremembered this! You do have the error on the chain and so if you loop the chain and print it then you get all of the errors, but if you just use the Display/tostring then you do not get the source printed out. On the other hand, if you add the source to the display but want the full chain and walk it, then you end up with sources printed out twice :(

**Comment by Hoverbear** [comment]
bors merge

**Comment by Hoverbear** [comment]
bors merge

**Comment by Hoverbear** [comment]
bors merge

**Pull request merged**